### PR TITLE
Initial work on stringless tokens

### PR DESCRIPTION
**Pull request by KristofferC**
Benchmark:

```julia
julia> @btime for t in tokenize(str, Tokens.RawToken); t; end;
  10.105 ms (89324 allocations: 4.09 MiB)

julia> @btime for t in tokenize(str, Tokens.Token); t; end;
  18.133 ms (243477 allocations: 10.22 MiB)
```

### CODE REVIEW

**Comment by KristofferC** [comment]
This is a bit bugged at the moment.. :P

**Comment by ZacLN** [comment]
I did a keyword/identifier [lexer](https://gist.github.com/ZacLN/977ce3e0be7d07f51ac45f538cf6c777)

its allocates nothing!

**Comment by KristofferC** [comment]
Haha, wow! Any changes in performance?

**Comment by ZacLN** [comment]
4-6 times faster, and it is ugly and may have bugs

**Comment by KristofferC** [comment]
Just curious, is the tokenizer a bottle neck in the parser code?

**Comment by KristofferC** [comment]
I wonder if that beast `lex_identifier` could be programatically generated from a set of keywords.

**Comment by ZacLN** [comment]
`next_token` takes up 50% of the time and  allocates 70% of the memory.

Yeah I thought about thhat but couldn't figure out how to do it quickly

**Comment by KristofferC** [comment]
Wow, I wouldn't have expected that. Should really optimize things here then!

**Comment by KristofferC** [comment]
I'm getting a lot of allocations in `done` when I run with `track-allocations` and `inline=no`... Not sure what that is about. If I run with inlining I can't really see at all where they happen (except in the `extract_tokenstring`).

**Comment by KristofferC** [comment]
We could keep the old version so it would be easy to switch back and forth to check for correctness.

**Comment by ZacLN** [comment]
Hmm, that sounds very strange I'll have a hunt around.

**Comment by ZacLN** [comment]
OK I'm not seeing allocations at`done` with `inline=no`, I'm going to try and merge my `lex_identifier` in and see if it works

**Comment by ZacLN** [comment]
Without checking for correctness this drops allocations by another half though `extract_token` is still being called somehow

```
# master 22ms 8.52mb
# stringless_tokens 17ms 4.33mb
# string_2 11ms 2.35mb
# string_2 7ms 208bytes (w/ reading of comments removed)
```

**Comment by KristofferC** [comment]
Why does reading comments allocate so much?

**Comment by ZacLN** [comment]
It wasn't the writing actually, its the tracking of line/column positions as well as start/end byte positions

```
# w/ line/col count
BenchmarkTools.Trial:
  memory estimate:  2.35 MiB
  allocs estimate:  51369
  --------------
  minimum time:     10.328 ms (0.00% GC)
  median time:      11.074 ms (0.00% GC)
  mean time:        11.395 ms (1.44% GC)
  maximum time:     20.929 ms (8.73% GC)
  --------------
  samples:          438
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

# wo/ line/col pos 
  BenchmarkTools.Trial:
  memory estimate:  208 bytes
  allocs estimate:  3
  --------------
  minimum time:     6.952 ms (0.00% GC)
  median time:      7.659 ms (0.00% GC)
  mean time:        7.640 ms (0.00% GC)
  maximum time:     9.012 ms (0.00% GC)
  --------------
  samples:          654
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

[stringless_2](https://github.com/ZacLN/Tokenize.jl/tree/stringless_2) fork for reference

**Comment by KristofferC** [comment]
I don't understand how that could allocate anything. Do you have a code diff between those two benchmarks?

**Comment by ZacLN** [comment]
I don't either: 
comment out line 265 of `lexer.jl`  in the fork linked above.

Timing used was:
```
using Tokenize, BenchmarkTools
function cyclethroughR(io)
    l = tokenize(io, Tokenize.Tokens.RawToken)
    while !Tokenize.Lexers.eof(l)
        Tokenize.Lexers.next_token(l)
    end 
end
str = readstring(Base.find_source_file("inference.jl"));
@benchmark cyclethroughR(IOBuffer(str))
```

**Comment by KristofferC** [comment]
Ah, but we still create the string object... Should easily be fixable

**Comment by KristofferC** [comment]
Something like that

**Comment by ZacLN** [comment]
Aye that cuts the timings from above to 8ms, ~~I may use another token type that doesn't store line/col to eke out that last 20% of speed~~

**Comment by KristofferC** [comment]
Can we get your `lex_identifier` function merged regardless of the stringless token stuff?

We could make `RawToken` store even less if some stuff are not needed in there.

**Comment by ZacLN** [comment]
It's only setup to work on `IOBuffer`s at the moment and I should refactor it a bit to use `emit` and the other existing structure properly.

Is there a reason to support both `Strings` and `IOBuffer` rather than just always convert to `IOBuffer`?

**Comment by KristofferC** [comment]
I benchmarked it and the timing seems very similar, (1.50s vs 1.57s). I think we should remove the string stuff then to make things simpler.

**Comment by ZacLN** [comment]
Ok then that makes things easier

**Comment by ZacLN** [comment]
This is failing to load? I've made the changes and they're ready to be inserted

**Comment by KristofferC** [comment]
Should be fixed now

**Comment by ZacLN** [comment]
Erg, I've added a PR to merge the new lex_identifier into this branch + some fixes

**Comment by KristofferC** [comment]
Rebased on master.

**Comment by KristofferC** [comment]
I added `untokenize` for `RawToken`. You need to give the `IO` or `String` that you originally used for the lexer and it extracts the string from the byte positions. It could be worth caching all keywords etc in a vector where each string is defined at the same index as the corresponding enum number. Then for all of those, extracting the string would just be an indexing into a vector.

**Comment by KristofferC** [comment]
Updated to use a lookup table for known strings. This took some manual work so it should be tested properly. `untokenizing` `inference.jl` changed from

```
 3.502 ms (89337 allocations: 5.41 MiB)
```
to
```
  2.775 ms (54825 allocations: 3.57 MiB)
```

**Comment by KristofferC** [comment]
Benchmarks for my machine with the code you previously benchmarked:

```jl
julia> @btime cyclethroughR(IOBuffer(str), Tokens.Token)
  11.348 ms (138752 allocations: 6.67 MiB)

julia> @btime cyclethroughR(IOBuffer(str), Tokens.RawToken)
  5.814 ms (2 allocations: 144 bytes)
```

Removing the `extract_tokenstring` call, i.e. don't keep track of rows and columns:

```jl
julia> @btime cyclethroughR(IOBuffer(str), Tokens.RawToken)
  4.007 ms (2 allocations: 144 bytes)
```

File: src/string_lut.jl [hunk 1]
```diff
@@ -80,7 +80,7 @@ NOT_IS               => "!==",
 IN                   => "in",
 ISA                  => "isa",
 LPIPE                => "|>",
-RPIPE                => "<",
+RPIPE                => "<|",
```
Comment (ZacLN) [inline src/string_lut.jl]: These names need to be switched/they parse as the opposing token kind

Comment (KristofferC) [reply src/string_lut.jl]: True!

Comment (ZacLN) [reply src/string_lut.jl]: Also in the comments of `token_kinds.jl`

Comment (KristofferC) [reply src/string_lut.jl]: Done

**Pull request closed without merging**